### PR TITLE
add MinMaxNorm constraint

### PR DESCRIPTION
**Pull request by cgratie**
Hi,

I have added a new constraint, for constraining the norm of the weights to an interval. In addition, the new constraint has a 'rate' parameter to control how the constraint is enforced.

All current constraints change the weights in a hard way to achieve the desired norm value. For example, consider the unit norm constraint. If we start training a network with random initial weights, the weights will already be rescaled to norm 1 after the first batch is processed. This might get the network stuck at a local minimum, because it moves the weights suddenly in a direction that might be incompatible with the gradient. The tunable 'rate' parameter allows the weights to change in a more predictable way, only doing a fragment of the rescaling required to reach norm 1 after each step. Hard constraint enforcing can still be achieved, by taking rate=1.0.

The MinMaxNorm constraint covers the previous ones as follows:
MaxNorm(m, axis) => MinMaxNorm(high=m, rate=1.0, axis=axis)
UnitNorm(axis) => MinMaxNorm(low=1.0, high=1.0, rate=1.0, axis=axis)

### CODE REVIEW

**Comment by fchollet** [comment]
Looks good to me. In theory we shouldn't change the API at this time, but this change is well-contained and makes sense so we may make an exception.

**Comment by bstriner** [comment]
How does this compare to using something like a boundary method to do constraints?

For example, regularization like `log(1-weight)+log(weight)` constrains weight to <0...1>, because the loss approaches infinity as weight approaches 0 or 1.

Anyone tested out the difference between the hard constraints we typically use and using a boundary method? Logarithmic boundary can give you NANs if the learning rate is too high, but there are other boundary methods possible.

Cheers

**Comment by cgratie** [comment]
One advantage I see for explicit interval constraints (hard or not) is that if the norm of the weights is already in the desired interval, no change is needed and learning goes in the direction of the gradient. 

This behavior can be achieved with boundary methods as well, but I find it less intuitive (I also find weight decay more intuitive than L2 regularization, so it might be just personal preference). For example, a quadratic loss term that would bring the weights closer to the `[a, b]` interval could be the following: `r * (max(a - weights, 0) ** 2 + max(weights - b, 0) ** 2)`. This is in fact exactly the kind of regularization that would happen when using something like `MinMaxConstraint(low=a, high=b, rate=r)` but for weight values instead of norms.

On the other hand, to achieve the regularization provided by hard constraints, one would need a boundary method that goes to infinity outside the desired interval (like your example), but more than that, it should be zero inside the interval. And as far as programming languages go, it is probably easier to write this as a hard constraint and avoid infinity.
